gadopt/stokes_integrators.py

```diff
@@ -169,13 +187,26 @@ def __init__(self, z, T, approximation, bcs=None, mu=1,
         self._solver_setup = False
 
     def setup_solver(self):
-        self.problem = fd.NonlinearVariationalProblem(self.F, self.solution,
-                                                      bcs=self.strong_bcs, J=self.J)
-        self.solver = fd.NonlinearVariationalSolver(self.problem,
-                                                    solver_parameters=self.solver_parameters,
-                                                    options_prefix=self.name,
-                                                    appctx={"mu": self.mu},
-                                                    **self.solver_kwargs)
+        if self.constant_jacobian:
+            z_tri = fd.TrialFunction(self.Z)
+            F_stokes_lin = fd.replace(self.F, {self.solution: z_tri})
+            a, L = fd.lhs(F_stokes_lin), fd.rhs(F_stokes_lin)
+            self.problem = fd.LinearVariationalProblem(a, L, self.solution,
```

Comment:
@stephankramer One of my motivations for overloading `J`, was that in `fd.LinearVariationalProbelm` it is assumed that `J = a` (`a` being the LHS). So if both `J` and `constant_jacobian` is provided, probably means that the user is not aware how this works? 

Additionally, what I still don't understand is the role of `aP` for preconditioning the system, and how this goes with all the other preconditioners we use. Our interface does not allow for providing `aP`.

Reply:
Ah yes, good point - it indeed doesn't make sense to set a J for the linear case, only makes sense in the context of a Newton solve - but I  do think the interface is cleaner now.

Hm yeah I can't immediately think of a good use case for Stokes (which is indeed probably why we don't provide an interface for it) - but say you have some weakly coupled ("mixed") linear system, then in `a` you would supply the full equations including the coupling terms, which correspond to off-diagonal blocks, but that system might be hard to solve directly, so you supply a `aP` where you leaves these terms out so you end up with a block-diagonal system and then you could set your preconditioner to `"pc_type": "ksp"` to exactly solve the block-diagonal system which would already be a good approximation as a preconditioner to the full system. So say in the future we want to couple Stokes and Temperature (also makes it nonlinear but let's ignore that) we could approximate it in `aP` with the uncoupled system. The equivalent in the nonlinear case would be Jp btw, so if we do add support for that, we probably just need to add `aP=Jp` as well for `constant_jacobian` case.

Reply:
Gotcha! As always thanks for the helpful explanation.